aten/src/ATen/native/cuda/UpSampleBilinear2d.cu

```diff
@@ -1,25 +1,312 @@
+  const int64_t width1 = idata.size(3);
+  const int64_t height2 = odata.size(2);
+  const int64_t width2 = odata.size(3);
+
+  if (index < n) {
+    const int64_t w2 = index % width2; // 0:width2-1
+    const int64_t h2 = index / width2; // 0:height2-1
+    // special case: just copy
+    if (height1 == height2 && width1 == width2) {
+      const int64_t h1 = h2;
+      const int64_t w1 = w2;
+      for (int n = 0; n < batchsize; n++) {
+        for (int c = 0; c < channels; ++c) {
+          const scalar_t val = idata[n][c][h1][w1];
+          odata[n][c][h2][w2] = val;
+        }
+      }
+      return;
+    }
+    //
+    const accscalar_t h1r = area_pixel_compute_source_index<accscalar_t>(
+        rheight, h2, align_corners, /*cubic=*/false);
+    const int64_t h1 = h1r;
+    const int64_t h1p = (h1 < height1 - 1) ? 1 : 0;
+    const accscalar_t h1lambda = h1r - h1;
+    const accscalar_t h0lambda = static_cast<accscalar_t>(1) - h1lambda;
+    //
+    const accscalar_t w1r = area_pixel_compute_source_index<accscalar_t>(
+        rwidth, w2, align_corners, /*cubic=*/false);
+    const int64_t w1 = w1r;
+    const int64_t w1p = (w1 < width1 - 1) ? 1 : 0;
+    const accscalar_t w1lambda = w1r - w1;
+    const accscalar_t w0lambda = static_cast<accscalar_t>(1) - w1lambda;
+    //
+    for (int n = 0; n < batchsize; n++) {
+      for (int c = 0; c < channels; ++c) {
+        const accscalar_t val = h0lambda *
+                (w0lambda * idata[n][c][h1][w1] +
+                 w1lambda * idata[n][c][h1][w1 + w1p]) +
+            h1lambda *
+                (w0lambda * idata[n][c][h1 + h1p][w1] +
+                 w1lambda * idata[n][c][h1 + h1p][w1 + w1p]);
+        odata[n][c][h2][w2] = static_cast<scalar_t>(val);
+      }
+    }
+  }
+}
+
+// Backward (adjoint) operation 1 <- 2 (accumulates)
+template <typename scalar_t, typename accscalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(1024)
+#endif
+__global__ void upsample_bilinear2d_backward_out_frame(
+    const int64_t n,
+    const accscalar_t rheight,
+    const accscalar_t rwidth,
+    const bool align_corners,
+    PackedTensorAccessor<scalar_t, 4> idata,
+    const PackedTensorAccessor<scalar_t, 4> odata) {
+  int64_t index = threadIdx.x + blockIdx.x * blockDim.x;
+
+  const int64_t batchsize = idata.size(0);
+  const int64_t channels = idata.size(1);
+  const int64_t height1 = idata.size(2);
+  const int64_t width1 = idata.size(3);
+  const int64_t height2 = odata.size(2);
+  const int64_t width2 = odata.size(3);
+
+  if (index < n) {
+    const int64_t w2 = index % width2; // 0:width2-1
+    const int64_t h2 = index / width2; // 0:height2-1
+    // special case: just copy
+    if (height1 == height2 && width1 == width2) {
+      const int64_t h1 = h2;
+      const int64_t w1 = w2;
+      for (int n = 0; n < batchsize; n++) {
+        for (int c = 0; c < channels; ++c) {
+          const scalar_t val = odata[n][c][h1][w1];
+          idata[n][c][h2][w2] += val;
```

Comment:
What's the contract here (and in all other *backward_out_* functions)? Is it accumulated into provided grad_input tensor, or filling grad_input with the result of interpolate_backward? Usually it's the latter for *_out functions, so `+=` is wrong?